db/memtable.cc

```diff
@@ -573,6 +590,13 @@ bool MemTable::Add(SequenceNumber s, ValueType type,
     }
   }
   if (type == kTypeRangeDeletion) {
+    fragmented_tombstones_mut.WriteLock();
+    if (s > last_range_del_seqno) {
```

Comment:
I think entries can be added to the memtable in non-sequence number order. @ajkr would know for certain.

Rather than checking sequence numbers, I think you can always invalidate the fragmented tombstones here by setting `fragmented_tombstones = nullptr`. Then in `Rebuild*` you can check to see if `fragmented_tombstones` is empty and rebuild. A final check is that you only want to update `fragmented_tombstones` in `Rebuild*` if the new set of fragmented tombstones is larger than the previous set. I'm fairly certain that the count of fragmented tombstones only ever increases as tombstones are added to the memtable. This prevents multiple concurrent rebuilds from stomping over each other and caching the wrong fragmented tombstones.

Reply:
Ah okay.  Thanks for the clarification here!